### PR TITLE
Add `set_bits` fuzz test

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

Part of https://github.com/apache/arrow-rs/pull/6288

# Rationale for this change
 
The code in https://github.com/apache/arrow-rs/pull/6288 is tricky and includes many special cases and `unsafe` so extra testing is warranted

# What changes are included in this PR?

Refactor the tests for `set_bits` into a struct and then implement a fuzzer that randomly generates test cases

# Are there any user-facing changes?
No, this is entirely testing

### CODE REVIEW

File: arrow-buffer/src/util/bit_mask.rs [hunk 1]
```diff
@@ -64,125 +64,235 @@ pub fn set_bits(
...
+            } else {
+                offset_read_bits / 8
+            };
+
+            // create space for writing
+            self.write_data.clear();
+            self.write_data
+                .resize(offset_write_bytes + len + extra_write_data_bytes, 0);
+
+            // interestingly set_bits seems to assume the output is already zeroed
```
Comment (alamb) [inline arrow-buffer/src/util/bit_mask.rs:252]: I was surprised at this -- if we fill the target with random data the tests fail (I am not sure why)

File: arrow-buffer/src/util/bit_mask.rs [hunk 2]
```diff
@@ -64,125 +64,235 @@ pub fn set_bits(
...
+            self.data.try_fill(rng).unwrap();
+            self.offset_read = offset_read_bits;
+
+            self.len = len;
+
+            // generated expectated output (not efficient)
+            self.expected_data.resize(self.write_data.len(), 0);
+            self.expected_data.copy_from_slice(&self.write_data);
+
+            self.expected_null_count = 0;
```
Comment (alamb) [inline arrow-buffer/src/util/bit_mask.rs:270]: The fuzz tester runs this code to compute the expected output (using `set_bit` and `get_bit`)

File: arrow-buffer/src/util/bit_mask.rs [hunk 3]
```diff
@@ -64,125 +64,235 @@ pub fn set_bits(
...
+                0b01111000, 0b01101001, 0b11100110, 0b11110110, 0b11111010, 0b11110000, 0b01111001,
+                0b01101001, 0b11100110, 0b11110110, 0b11111010, 0b11110000, 0b00000001,
+            ],
+            expected_null_count: 35,
+        }
+        .verify();
+    }
+
+    #[test]
+    fn set_bits_fuz() {
```
Comment (alamb) [inline arrow-buffer/src/util/bit_mask.rs:154]:
```suggestion
    fn set_bits_fuzz() {
```

File: arrow-buffer/src/util/bit_mask.rs [hunk 4]
```diff
@@ -64,125 +64,235 @@ pub fn set_bits(
...
+        let mut rng = StdRng::seed_from_u64(42);
+        let mut data = SetBitsTest::new();
+        for _ in 0..10000 {
```
Comment (alamb) [inline arrow-buffer/src/util/bit_mask.rs:157]: 10000 takes too long, lets use less

```suggestion
        for _ in 0..100 {
```

**Review by kazuyukitanimura — APPROVED** [review]
Thank you @alamb for creating this

**Review by Dandandan — APPROVED** [review]

**Comment by alamb** [comment]
Thank you @Dandandan

**Pull request merged**